pkg/controllers/cronfederatedhpa/cronfederatedhpa_controller.go

```
@@ -0,0 +1,204 @@
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	autoscalingv1alpha1 "github.com/karmada-io/karmada/pkg/apis/autoscaling/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+const (
+	// ControllerName is the controller name that will be used when reporting events.
+	ControllerName = "cronfederatedhpa-controller"
+)
+
+// CronFHPAController is used to operate CronFederatedHPA.
+type CronFHPAController struct {
+	client.Client // used to operate Cron resources.
+	EventRecorder record.EventRecorder
+
+	RateLimiterOptions ratelimiterflag.Options
+	CronHandler        *CronHandler
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *CronFHPAController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling CronFederatedHPA %s", req.NamespacedName)
+
+	cronFHPA := &autoscalingv1alpha1.CronFederatedHPA{}
+	if err := c.Client.Get(ctx, req.NamespacedName, cronFHPA); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof("Begin to cleanup the cron jobs for CronFederatedHPA:%s", req.NamespacedName)
+			c.CronHandler.StopCronFHPAExecutor(req.NamespacedName.String())
+			return controllerruntime.Result{}, nil
+		}
+
+		klog.Error("Fail to get CronFederatedHPA(%s):", req.NamespacedName, err)
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			c.EventRecorder.Event(cronFHPA, corev1.EventTypeWarning, "ProcessCronFederatedHPAFailed", err.Error())
+		}
+	}()
```

It would be better to report the event immediately once found an error.
So that, we can report a concrete error instead of a general one.